runners/spark/src/main/java/org/apache/beam/runners/spark/aggregators/AccumulatorSingleton.java

```diff
@@ -19,35 +19,114 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 
 /**
  * For resilience, {@link Accumulator}s are required to be wrapped in a Singleton.
  * @see <a href="https://spark.apache.org/docs/1.6.3/streaming-programming-guide.html#accumulators-and-broadcast-variables">accumulators</a>
  */
-class AccumulatorSingleton {
+public class AccumulatorSingleton {
+  private static final Logger LOG = LoggerFactory.getLogger(AccumulatorSingleton.class);
+
+  private static final String ACCUMULATOR_CHECKPOINT_FILE = "beam_aggregators";
 
-  private static volatile Accumulator<NamedAggregators> instance = null;
+  private static volatile Accumulator<NamedAggregators> instance;
+  private static volatile FileSystem fileSystem;
+  private static volatile Path checkpointPath;
+  private static volatile Path tempCheckpointPath;
+  private static volatile Path backupCheckpointPath;
 
-  static Accumulator<NamedAggregators> getInstance(JavaSparkContext jsc) {
+  static Accumulator<NamedAggregators> getInstance(
+      JavaSparkContext jsc,
+      boolean isStreaming,
+      String checkpointDir) {
     if (instance == null) {
       synchronized (AccumulatorSingleton.class) {
         if (instance == null) {
-          //TODO: currently when recovering from checkpoint, Spark does not recover the
-          // last known Accumulator value. The SparkRunner should be able to persist and recover
-          // the NamedAggregators in order to recover Aggregators as well.
           instance = jsc.sc().accumulator(new NamedAggregators(), new AggAccumParam());
+          if (isStreaming) {
+            recoverValueFromCheckpoint(jsc, checkpointDir);
+          }
         }
       }
     }
     return instance;
   }
 
+  private static void recoverValueFromCheckpoint(JavaSparkContext jsc, String checkpointDir) {
+    FSDataInputStream is = null;
+    try {
+      checkpointPath = new Path(checkpointDir, ACCUMULATOR_CHECKPOINT_FILE);
+      tempCheckpointPath = checkpointPath.suffix(".tmp");
+      backupCheckpointPath = checkpointPath.suffix(".bak");
+      fileSystem = checkpointPath.getFileSystem(jsc.hadoopConfiguration());
+      if (fileSystem.exists(checkpointPath)) {
+        is = fileSystem.open(checkpointPath);
+      } else if (fileSystem.exists(backupCheckpointPath)) {
+        is = fileSystem.open(backupCheckpointPath);
+      }
+      if (is != null) {
+        ObjectInputStream objectInputStream = new ObjectInputStream(is);
+        NamedAggregators recoveredValue =
+            (NamedAggregators) objectInputStream.readObject();
+        objectInputStream.close();
+        LOG.info("Recovered accumulators from checkpoint: " + recoveredValue);
+        instance.setValue(recoveredValue);
+      } else {
+        LOG.info("No accumulator checkpoint found.");
+      }
+    } catch (Exception e) {
+      throw new RuntimeException("Failure while reading accumulator checkpoint.", e);
+    }
+  }
+
+  private static void checkpoint() throws IOException {
```

Comment:
I'd prefer not to write into Spark's checkpoint dir, maybe create a runner checkpoint dir with Spark's own as sub-dir, and the accumulator state there.